test/lib/users.js

```diff
@@ -68,56 +68,26 @@ describe("lib/users", function() {
 	});
 
 	describe("class User", function() {
-		let roles = new Map([
-			[1, new lib.Role({ id: 1, name: "a", description: "a role", permissions: ["core.admin"] })],
-			[2, new lib.Role({ id: 2, name: "b", description: "b role", permissions: ["test"] })],
-		]);
 		it("should round trip serialize", function() {
 			function test_roundtrip(serialized) {
-				let user = new lib.User(serialized, roles);
-				let user_serialized = user.serialize();
+				let user = lib.User.fromJSON(serialized);
+				let user_serialized = user.toJSON();
 				assert.deepEqual(user_serialized, serialized);
-				let user_deserialized = new lib.User(user_serialized, roles);
+				let user_deserialized = lib.User.fromJSON(user_serialized);
 				assert.deepEqual(user_deserialized, user);
 			}
 
 			test_roundtrip({ name: "admin", roles: [1] });
-			test_roundtrip({ name: "user", roles: [2], token_valid_after: 12345 });
 			test_roundtrip({ name: "user", is_admin: true, is_whitelisted: true });
 			test_roundtrip({ name: "user", is_banned: true, ban_reason: "Bad user" });
 			test_roundtrip({ name: "user", instance_stats: [[1, { join_count: 1 }]]});
 		});
-		it("should ignore invalid roles", function() {
-			let user = new lib.User({ name: "test", roles: [1, 4, 55] }, roles);
-			assert.equal(user.roles.size, 1, "Unexpected count of roles");
-		});
-		it("should track online users", function() {
-			let user = new lib.User({ name: "admin", roles: [1] }, roles);
-			assert(!lib.User.onlineUsers.has(user));
-			assert.deepEqual(user.instances, new Set());
-
-			user.notifyJoin(12);
-			assert(lib.User.onlineUsers.has(user));
-			assert.deepEqual(user.instances, new Set([12]));
-
-			user.notifyJoin(8);
-			assert(lib.User.onlineUsers.has(user));
-			assert.deepEqual(user.instances, new Set([12, 8]));
-
-			user.notifyLeave(11);
-			assert(lib.User.onlineUsers.has(user));
-			assert.deepEqual(user.instances, new Set([12, 8]));
-
-			user.notifyLeave(12);
-			assert(lib.User.onlineUsers.has(user));
-			assert.deepEqual(user.instances, new Set([8]));
-
-			user.notifyLeave(8);
-			assert(!lib.User.onlineUsers.has(user));
-			assert.deepEqual(user.instances, new Set());
+		it("should keep invalid roles", function() {
+			let user = lib.User.fromJSON({ name: "test", roles: [1, 4, 55] });
+			assert.equal(user.roleIds.size, 3, "Unexpected count of roles");
```

Comment:
Why is this desired?

Reply:
Because user is a data class that is indpendent from role now so it shouldn't filter invalid roles and because it can't access the roles that exist it can't filter them either.